plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/data/MapboxViewportDataSource.kt

```diff
@@ -0,0 +1,272 @@
+      val pointsForFollowing = listOf(nonNullTargetLocation.toPoint())
+
+      options.followingFrameOptions.bearingSmoothing.run {
+        val locationBearing = nonNullTargetLocation.bearing.toDouble() ?: BEARING_NORTH
+        followingBearingProperty.fallback =
+          getSmootherBearingForMap(enabled, cameraManager.cameraState.bearing, locationBearing)
+      }
+
+      val cameraFrame = if (pointsForFollowing.size > 1 &&
+        options.followingFrameOptions.maximizeViewableGeometryWhenPitchZero &&
+        followingPitchProperty.get() == ZERO_PITCH
+      ) {
+        cameraManager.cameraForCoordinates(
+          pointsForFollowing,
+          followingPadding,
+          followingBearingProperty.get(),
+          followingPitchProperty.get()
+        )
+      } else {
+        val mapSize = mapTransform.getSize()
+        val screenBox = getScreenBoxForFraming(mapSize, followingPadding)
+        val cameraState = cameraManager.cameraState
+        val fallbackCameraOptions = CameraOptions.Builder()
+          .center(pointsForFollowing.firstOrNull() ?: cameraState.center)
+          .padding(getMapAnchoredPaddingFromUserPadding(mapSize, followingPadding))
+          .bearing(followingBearingProperty.get())
+          .pitch(followingPitchProperty.get())
+          .zoom(cameraState.zoom)
+          .build()
+        if (pointsForFollowing.size > 1) {
+          cameraManager.cameraForCoordinates(
+            pointsForFollowing,
+            fallbackCameraOptions,
+            screenBox
+          )
+        } else {
+          fallbackCameraOptions
+        }
+      }
+
+      followingCenterProperty.fallback = cameraFrame.center!!
+      options.followingFrameOptions.run {
+        followingZoomProperty.fallback = max(min(cameraFrame.zoom!!, maxZoom), minZoom)
+      }
+      appliedFollowingPadding = cameraFrame.padding!!
+      return
+    }
+    // nothing to frame
+    options.followingFrameOptions.run {
+      val cameraState = cameraManager.cameraState
+      followingBearingProperty.fallback = cameraState.bearing
+      followingPitchProperty.fallback = defaultPitch
+      followingCenterProperty.fallback = cameraState.center
+      followingZoomProperty.fallback = max(min(cameraState.zoom, maxZoom), minZoom)
+    }
+  }
+
+  private fun updateOverviewData() {
+    targetLocation?.let { nonNullTargetLocation ->
+      val pointsForOverview = listOf(nonNullTargetLocation.toPoint())
+
+      overviewBearingProperty.fallback = normalizeBearing(
+        cameraManager.cameraState.bearing,
+        BEARING_NORTH
+      )
+
+      val cameraFrame = if (pointsForOverview.isNotEmpty()) {
+        cameraManager.cameraForCoordinates(
+          pointsForOverview,
+          overviewPadding,
+          overviewBearingProperty.get(),
+          overviewPitchProperty.get()
+        )
+      } else {
+        cameraManager.cameraState.toCameraOptions()
+      }
+
+      overviewCenterProperty.fallback = cameraFrame.center!!
```

Comment:
I'd avoid using `!!` inside SDK files, let's handle those errors somehow (use `RuntimeException` with some text). Applies to other things below

Reply:
Will revisit it in the follow up PR to refactor the `MapboxViewportDataSource`